File: src/adapters/network.ts
import axios, { type AxiosRequestConfig } from "axios";
import { HttpProxyAgent, HttpsProxyAgent } from "hpagent";
import { SocksProxyAgent } from "socks-proxy-agent";
import { consola } from "consola";
import type { AppConfig } from "../config/index.js";
import type { FetchOptions } from "../core/types.js";

// ==============================================================================
// 智能网络分流与多协议抓取适配器 (Smart Routing Network Adapter)
// ==============================================================================
export class NetworkAdapter {
  private httpAgent?: any;
  private httpsAgent?: any;

  constructor(private config: AppConfig["network"]) {
    this.initProxyAgents();
  }

  private initProxyAgents(): void {
    if (!this.config.proxyUrl || this.config.proxyMode === "never") {
      return;
    }

    try {
      const proxyUrl = new URL(this.config.proxyUrl);
      if (this.config.proxyAuth && !proxyUrl.username) {
        const [u, p] = this.config.proxyAuth.split(":");
        proxyUrl.username = u;
        proxyUrl.password = p;
      }

      if (proxyUrl.protocol.startsWith("socks")) {
        const agent = new SocksProxyAgent(proxyUrl.toString());
        this.httpAgent = agent;
        this.httpsAgent = agent;
      } else {
        this.httpAgent = new HttpProxyAgent({
          keepAlive: true,
          keepAliveMsecs: 1000,
          maxSockets: 256,
          maxFreeSockets: 256,
          proxy: proxyUrl.toString(),
        });
        this.httpsAgent = new HttpsProxyAgent({
          keepAlive: true,
          keepAliveMsecs: 1000,
          maxSockets: 256,
          maxFreeSockets: 256,
          proxy: proxyUrl.toString(),
        });
      }
    } catch (err) {
      consola.warn("⚠️ 初始化代理 Agent 失败，将降级为直连:", err);
    }
  }

  /**
   * 判断目标 URL 是否属于国内直连范围 (智能分流白名单)
   */
  isDomesticUrl(urlStr: string): boolean {
    try {
      const host = new URL(urlStr).hostname.toLowerCase();
      return this.config.domesticDomains.some((keyword) => {
        if (keyword.startsWith(".")) {
          return host.endsWith(keyword);
        }
        return host === keyword || host.endsWith(`.${keyword}`);
      });
    } catch {
      return false;
    }
  }

  /**
   * 执行 HTTP GET 抓取，智能挂载代理或直连
   */
  async fetch(url: string, options?: FetchOptions): Promise<string> {
    const isDomestic =
      options?.isDomestic ?? (this.config.proxyMode === "auto" ? this.isDomesticUrl(url) : false);

    const useProxy =
      this.config.proxyMode === "always" ||
      (this.config.proxyMode === "auto" && !isDomestic && !!this.httpAgent);

    const axiosConfig: AxiosRequestConfig = {
      url,
      method: "GET",
      timeout: options?.timeoutMs || this.config.timeoutMs,
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
        Accept:
          "text/html,application/xhtml+xml,application/xml;q=0.9,application/json;q=0.8,*/*;q=0.7",
        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
        ...options?.headers,
      },
      responseType: options?.responseType === "json" ? "json" : "text",
    };

    if (useProxy) {
      axiosConfig.httpAgent = this.httpAgent;
      axiosConfig.httpsAgent = this.httpsAgent;
      axiosConfig.proxy = false; // 由 hpagent/socks-proxy-agent 托管
    } else {
      axiosConfig.proxy = false;
    }

    try {
      const res = await axios(axiosConfig);
      return typeof res.data === "string" ? res.data : JSON.stringify(res.data);
    } catch (err: any) {
      const routeInfo = useProxy ? `[代理: ${this.config.proxyUrl}]` : `[直连]`;
      throw new Error(`网络请求失败 ${routeInfo} (${url}): ${err.message || err}`);
    }
  }

  /**
   * 专用的 JSON 抓取快捷方法
   */
  async fetchJson<T = any>(url: string, options?: FetchOptions): Promise<T> {
    const raw = await this.fetch(url, { ...options, responseType: "json" });
    try {
      return typeof raw === "string" ? JSON.parse(raw) : raw;
    } catch {
      return raw as any;
    }
  }
}

File: utils/network.py
"""
智能网络分流客户端 (Smart Network Dispatcher)
严格遵循：
  - 国内网站绝对直连 (Direct Connection)，防止 WAF 拦截与节点绕路
  - 国外网站自适应走本地宿主机 Clash 代理 (7897端口隧道)
"""
import urllib.request
import ssl
import socket
import os
import sys

from config.settings import (
    IS_WSL, CLASH_PORT, CLASH_HOST_WINDOWS, CLASH_HOST_WSL, 
    CLASH_AUTH, REQUEST_TIMEOUT, DOMESTIC_KEYWORDS
)

class SmartNetworkClient:
    def __init__(self):
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE

        # 1. 国内直连通道 (Direct Opener)
        self.direct_opener = urllib.request.build_opener(
            urllib.request.ProxyHandler({}),
            urllib.request.HTTPSHandler(context=ctx)
        )

        # 2. 国外代理通道 (Proxy Opener)
        # 判断运行环境并拼接代理地址
        if IS_WSL:
            proxy_url = f"http://{CLASH_AUTH}@{CLASH_HOST_WSL}:{CLASH_PORT}"
        else:
            proxy_url = f"http://{CLASH_HOST_WINDOWS}:{CLASH_PORT}"

        self.proxy_opener = urllib.request.build_opener(
            urllib.request.ProxyHandler({
                "http": proxy_url,
                "https": proxy_url
            }),
            urllib.request.HTTPSHandler(context=ctx)
        )

    def is_domestic(self, url: str) -> bool:
        url_lower = url.lower()
        return any(kw in url_lower for kw in DOMESTIC_KEYWORDS)

    def fetch(self, url: str, headers: dict = None, timeout: int = REQUEST_TIMEOUT) -> bytes:
        default_headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36",
            "Accept": "text/html,application/xhtml+xml,application/xml,application/json;q=0.9,*/*;q=0.8"
        }
        if headers:
            default_headers.update(headers)

        req = urllib.request.Request(url, headers=default_headers)

        # 智能分流
        if self.is_domestic(url):
            opener = self.direct_opener
            mode = "Direct (国内直连)"
        else:
            opener = self.proxy_opener
            mode = "Clash Proxy (国外代理)"

        try:
            with opener.open(req, timeout=timeout) as resp:
                data = resp.read()
                return data
        except Exception as e:
            # 若代理异常，国外源尝试一次直连回退
            if not self.is_domestic(url):
                try:
                    with self.direct_opener.open(req, timeout=timeout) as resp:
                        return resp.read()
                except Exception:
                    pass
            raise e

# 全局单例
client = SmartNetworkClient()
